src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -0,0 +1,526 @@
+        FileDigest,
+        ExtractFileDigest(stripped_digest, file_path.name),
+    )
+    if file_digest != request.file_digest:
+        raise CoursierError(
+            f"Coursier fetch for '{resolved_coord}' succeeded, but fetched artifact {file_digest} did not match the expected artifact: {request.file_digest}."
+        )
+    return ResolvedClasspathEntry(
+        coord=request.coord, file_name=file_path.name, digest=stripped_digest
+    )
+
+
+@rule(level=LogLevel.DEBUG)
+async def coursier_fetch_lockfile(lockfile: CoursierResolvedLockfile) -> ResolvedClasspathEntries:
+    """Fetch every artifact in a lockfile."""
+    classpath_entries = await MultiGet(
+        Get(ResolvedClasspathEntry, CoursierLockfileEntry, entry) for entry in lockfile.entries
+    )
+    return ResolvedClasspathEntries(classpath_entries)
+
+
+@rule(level=LogLevel.DEBUG)
+async def load_coursier_lockfile_from_source(
+    lockfile_field: JvmLockfileSources,
+) -> CoursierResolvedLockfile:
+    lockfile_sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            [lockfile_field],
+            for_sources_types=[JvmLockfileSources],
+            enable_codegen=False,
+        ),
+    )
+    if len(lockfile_sources.files) != 1:
+        raise CoursierError("JvmLockfileSources must have exactly 1 source file")
+
+    source_lockfile_digest_contents = await Get(
+        DigestContents, Digest, lockfile_sources.snapshot.digest
+    )
+    source_lockfile_content = source_lockfile_digest_contents[0]
+    return CoursierResolvedLockfile.from_json_dict(json.loads(source_lockfile_content.content))
+
+
+@dataclass(frozen=True)
+class CoursierLockfileForTargetRequest:
+    targets: Targets
+
+
+@rule
+async def get_coursier_lockfile_for_target(
+    request: CoursierLockfileForTargetRequest,
+) -> CoursierResolvedLockfile:
+    """Determine the lockfile that applies to a given JVM target.
+
+    Presently this just walks the target's transitive dependencies to find a dependency
+    that provides `JvmLockfileSources`.
+
+    This rule asserts that at most 1 such dependency may be in the JVM target's transitive
+    dependencies.
+    """
+
+    transitive_targets = await Get(
+        TransitiveTargets, TransitiveTargetsRequest(target.address for target in request.targets)
+    )
+    transitive_jvm_lockfile_sources = [
+        target[JvmLockfileSources]
+        for target in transitive_targets.closure
+        if target.has_field(JvmLockfileSources)
+    ]
+    if len(transitive_jvm_lockfile_sources) == 0:
+        raise CoursierError(
+            "Exactly 1 target with a coursier_lockfile should appear in the transitive closure of a JVM target, but none were found."
+        )
+
+    if len(transitive_jvm_lockfile_sources) > 1:
+        raise CoursierError(
+            f"Exactly 1 target with a coursier_lockfile should appear in the transitive closure of a JVM library, but {len(transitive_jvm_lockfile_sources)} were found."
```

Comment:
Looks like this is >100 lines. Black won't format this, so you're encouraged to use soft wrapping to split over multiple lines.